File: 2016/day05_2.py
# pylint: disable=missing-module-docstring, missing-function-docstring


import hashlib
from typing import Any


def solution(content: str) -> Any:
    password = [""] * 8
    generator = hash_gen(content)
    while len([p for p in password if p]) < 8:
        hashed = next(generator)
        index = int(hashed[5], 16)
        if index > 7:
            continue
        if password[index]:
            continue
        password[index] = hashed[6]
    return "".join(password)


def hash_gen(prefix: str):
    counter = 0
    while True:
        hashed = hashlib.md5(bytes(prefix + str(counter), "utf8")).hexdigest()
        if hashed.startswith("0" * 5):
            yield hashed
        counter += 1


def main() -> None:
    with open("day05.dat", "r", encoding="utf8") as file:
        print(solution(file.read().strip()))


if __name__ == "__main__":
    main()

File: 2016/day05-2.ts
import { encodeHex } from "jsr:@std/encoding/hex";
import { md5 } from "jsr:@takker/md5";

function solution(content: string) {
  const g = generate(content);
  const password = Array.from({ length: 8 }).map((_) => "");
  while (password.filter((it) => it).length < 8) {
    const hashed = g.next().value;
    const index = parseInt(hashed[5], 16);
    if (index > 7) {
      continue;
    }
    if (password[index]) {
      continue;
    }
    password[index] = hashed[6];
  }
  return password.join("");
}

function* generate(prefix: string): Generator<string, string, unknown> {
  let counter = 0;
  while (true) {
    const hashed = encodeHex(md5(prefix + counter));
    if (hashed.startsWith("00000")) {
      yield hashed;
    }
    counter += 1;
  }
}

Deno.readTextFile("day05.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
